# Importing Necessary Files

In [89]:
import pandas as pd
import numpy as np

import seaborn as sns
import matplotlib.pyplot as plt

import warnings
warnings.filterwarnings('ignore')

# Importing Data files

In [90]:
Address_test = "https://datahack-prod.s3.amazonaws.com/test_file/test_AbJTz2l.csv"
test_df = pd.read_csv(Address_test)

In [91]:
Address_train = 'https://datahack-prod.s3.amazonaws.com/train_file/train_v9rqX0R.csv'
train_df = pd.read_csv(Address_train)

# Data Analysis and Preprocessing

In [92]:
test_df.head()

Item_Identifier  Item_Weight Item_Fat_Content  Item_Visibility    Item_Type  \
0           FDW58       20.750          Low Fat         0.007565  Snack Foods   
1           FDW14        8.300              reg         0.038428        Dairy   
2           NCN55       14.600          Low Fat         0.099575       Others   
3           FDQ58        7.315          Low Fat         0.015388  Snack Foods   
4           FDY38          NaN          Regular         0.118599        Dairy   

   Item_MRP Outlet_Identifier  Outlet_Establishment_Year Outlet_Size  \
0  107.8622            OUT049                       1999      Medium   
1   87.3198            OUT017                       2007         NaN   
2  241.7538            OUT010                       1998         NaN   
3  155.0340            OUT017                       2007         NaN   
4  234.2300            OUT027                       1985      Medium   

  Outlet_Location_Type        Outlet_Type  
0               Tier 1  Supermarket Type1  
1               Tier 2  Supermarket Type1  
2               Tier 3      Grocery Store  
3               Tier 2  Supermarket Type1  
4               Tier 3  Supermarket Type3

In [93]:
train_df.head()

Item_Identifier  Item_Weight Item_Fat_Content  Item_Visibility  \
0           FDA15         9.30          Low Fat         0.016047   
1           DRC01         5.92          Regular         0.019278   
2           FDN15        17.50          Low Fat         0.016760   
3           FDX07        19.20          Regular         0.000000   
4           NCD19         8.93          Low Fat         0.000000   

               Item_Type  Item_MRP Outlet_Identifier  \
0                  Dairy  249.8092            OUT049   
1            Soft Drinks   48.2692            OUT018   
2                   Meat  141.6180            OUT049   
3  Fruits and Vegetables  182.0950            OUT010   
4              Household   53.8614            OUT013   

   Outlet_Establishment_Year Outlet_Size Outlet_Location_Type  \
0                       1999      Medium               Tier 1   
1                       2009      Medium               Tier 3   
2                       1999      Medium               Tier 1   
3                       1998         NaN               Tier 3   
4                       1987        High               Tier 3   

         Outlet_Type  Item_Outlet_Sales  
0  Supermarket Type1          3735.1380  
1  Supermarket Type2           443.4228  
2  Supermarket Type1          2097.2700  
3      Grocery Store           732.3800  
4  Supermarket Type1           994.7052

In [94]:
test_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5681 entries, 0 to 5680
Data columns (total 11 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   Item_Identifier            5681 non-null   object 
 1   Item_Weight                4705 non-null   float64
 2   Item_Fat_Content           5681 non-null   object 
 3   Item_Visibility            5681 non-null   float64
 4   Item_Type                  5681 non-null   object 
 5   Item_MRP                   5681 non-null   float64
 6   Outlet_Identifier          5681 non-null   object 
 7   Outlet_Establishment_Year  5681 non-null   int64  
 8   Outlet_Size                4075 non-null   object 
 9   Outlet_Location_Type       5681 non-null   object 
 10  Outlet_Type                5681 non-null   object 
dtypes: float64(3), int64(1), object(7)
memory usage: 488.3+ KB


In [95]:
train_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8523 entries, 0 to 8522
Data columns (total 12 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   Item_Identifier            8523 non-null   object 
 1   Item_Weight                7060 non-null   float64
 2   Item_Fat_Content           8523 non-null   object 
 3   Item_Visibility            8523 non-null   float64
 4   Item_Type                  8523 non-null   object 
 5   Item_MRP                   8523 non-null   float64
 6   Outlet_Identifier          8523 non-null   object 
 7   Outlet_Establishment_Year  8523 non-null   int64  
 8   Outlet_Size                6113 non-null   object 
 9   Outlet_Location_Type       8523 non-null   object 
 10  Outlet_Type                8523 non-null   object 
 11  Item_Outlet_Sales          8523 non-null   float64
dtypes: float64(4), int64(1), object(7)
memory usage: 799.2+ KB


In [96]:
train_df.shape

(8523, 12)

In [97]:
test_df.shape

(5681, 11)

In [98]:
#Column Names of test Data
test_df.columns

Index(['Item_Identifier', 'Item_Weight', 'Item_Fat_Content', 'Item_Visibility',
       'Item_Type', 'Item_MRP', 'Outlet_Identifier',
       'Outlet_Establishment_Year', 'Outlet_Size', 'Outlet_Location_Type',
       'Outlet_Type'],
      dtype='object')

In [99]:
#Column Names of Train Data
train_df.columns

Index(['Item_Identifier', 'Item_Weight', 'Item_Fat_Content', 'Item_Visibility',
       'Item_Type', 'Item_MRP', 'Outlet_Identifier',
       'Outlet_Establishment_Year', 'Outlet_Size', 'Outlet_Location_Type',
       'Outlet_Type', 'Item_Outlet_Sales'],
      dtype='object')

In [100]:
#Finding Null values in Training Data

for column_name in train_df.columns:
    count = train_df[column_name][train_df[column_name].isnull()].size
    print(column_name,count)

Item_Identifier 0
Item_Weight 1463
Item_Fat_Content 0
Item_Visibility 0
Item_Type 0
Item_MRP 0
Outlet_Identifier 0
Outlet_Establishment_Year 0
Outlet_Size 2410
Outlet_Location_Type 0
Outlet_Type 0
Item_Outlet_Sales 0


In [101]:
#Finding Unique Values in each Column

for column_name in train_df.columns:
  print(column_name, train_df[column_name].unique())

Item_Identifier ['FDA15' 'DRC01' 'FDN15' ... 'NCF55' 'NCW30' 'NCW05']
Item_Weight [ 9.3    5.92  17.5   19.2    8.93  10.395 13.65     nan 16.2   11.8
 18.5   15.1   17.6   16.35   9.    13.35  18.85  14.6   13.85  13.
  7.645 11.65   5.925 19.25  18.6   18.7   17.85  10.     8.85   9.8
 13.6   21.35  12.15   6.42  19.6   15.85   7.39  10.195  9.895 10.895
  7.905  9.195  8.365  7.97  17.7   19.35   8.645 15.6   18.25   7.855
  7.825  8.39  12.85  19.     5.905  7.76  16.75  15.5    6.055  6.305
 20.85  20.75   8.895 19.7    8.75  13.3    8.31  19.75  17.1   10.5
  6.635 14.15   8.89   9.1    7.5   16.85   7.485 11.6   12.65  20.25
  8.6   12.6    8.88  20.5   13.5    7.235  6.92   8.02  12.8   16.6
 14.    16.    21.25   7.365 18.35   5.465  7.27   6.155 19.5   15.2
 14.5   13.1   12.3   11.1   11.3    5.75  11.35   6.525 10.3    5.78
 11.85  18.75   5.26  16.1    9.5   13.8   14.65   6.67   6.11  17.2
  6.32   4.88   5.425 14.1    7.55  17.25  12.    10.1    7.785 13.15
  8.5    7.63

In [102]:
#Replacing Null Values in "Item_Weight" with 0

train_df.loc[train_df.Item_Weight.isnull(), "Item_Weight"]=0

In [103]:
#Replacing Null Values in "Outlet_Size" with Undefined

train_df.loc[train_df.Outlet_Size.isnull(), "Outlet_Size"]="Undefined"

In [104]:
#Finding Null values in Test Data

for column_name in test_df.columns:
    count = test_df[column_name][test_df[column_name].isnull()].size
    print(column_name,count)

Item_Identifier 0
Item_Weight 976
Item_Fat_Content 0
Item_Visibility 0
Item_Type 0
Item_MRP 0
Outlet_Identifier 0
Outlet_Establishment_Year 0
Outlet_Size 1606
Outlet_Location_Type 0
Outlet_Type 0


In [105]:
#Finding Unique Values in each Column

for column_name in test_df.columns:
  print(column_name, test_df[column_name].unique())

Item_Identifier ['FDW58' 'FDW14' 'NCN55' ... 'NCI29' 'FDP28' 'FDF04']
Item_Weight [20.75   8.3   14.6    7.315    nan  9.8   19.35   6.305  5.985 16.6
  6.59   4.785 16.75   6.135 19.85  17.85  13.6    7.1   19.2   13.65
  6.785 13.     5.825 13.8   12.8    9.5   12.5    5.78  10.395 18.35
 21.1   19.75   8.43  15.35   6.71  10.195 11.65  12.85  12.15  18.25
  5.98  12.65  13.1   20.7    7.855 13.35  17.25   8.315 21.35   5.88
  7.435 14.65  14.    18.2   17.6    7.71   7.81  16.35  16.1    5.34
  5.51   6.61   6.38   8.71  17.75   6.44  12.6    7.935 18.5    6.035
 11.8    6.26  16.5   15.7    6.635  8.155  6.175  8.18   8.355  6.905
 20.35  19.25   5.48   5.425 12.1   11.     5.765  5.44   6.13   8.195
 14.15  14.3    5.635  5.73   7.155 15.2   15.25   8.76   9.13  19.6
 10.    11.3    6.57  13.15   4.635 10.6    9.3    7.02   7.825 19.1
  7.51   8.85  17.35  15.15   5.46  13.5    8.275 10.8   15.5    4.905
  9.285  8.97  18.75   8.895  6.89  16.25  13.85  18.     9.21  20.85
 20.25 

In [106]:
#Replacing Null Values in "Item_Weight" with 0

test_df.loc[test_df.Item_Weight.isnull(), "Item_Weight"]=0

In [107]:
#Replacing Null Values in "Outlet_Size" with Undefined

test_df.loc[test_df.Outlet_Size.isnull(), "Outlet_Size"]="Undefined"

In [108]:
# Item_Fat_Content column contains duplicate values. Changing it to unique values

train_df.loc[train_df.Item_Fat_Content=='Low Fat', "Item_Fat_Content"] = "LF"
train_df.loc[train_df.Item_Fat_Content=='low fat', "Item_Fat_Content"] = "LF"
train_df.loc[train_df.Item_Fat_Content=='reg', "Item_Fat_Content"] = "Reg"

In [109]:
train_df.Item_Fat_Content.value_counts()

LF         5517
Regular    2889
Reg         117
Name: Item_Fat_Content, dtype: int64

In [110]:
# Item_Fat_Content column contains duplicate values. Changing it to unique values

test_df.loc[test_df.Item_Fat_Content=='Low Fat', "Item_Fat_Content"] = "LF"
test_df.loc[test_df.Item_Fat_Content=='low fat', "Item_Fat_Content"] = "LF"
test_df.loc[test_df.Item_Fat_Content=='reg', "Item_Fat_Content"] = "Reg"

In [111]:
test_df.Item_Fat_Content.value_counts()

LF         3668
Regular    1935
Reg          78
Name: Item_Fat_Content, dtype: int64

In [112]:
#Deleting the columns having Item ID and Outlet ID from both Train Data

train_df.drop('Item_Identifier', axis = 1, inplace = True)
train_df.drop('Outlet_Identifier',axis = 1, inplace = True)
#test_df.drop('Item_Identifier', axis = 1, inplace = True)
#test_df.drop('Outlet_Identifier',axis = 1, inplace = True)


In [49]:
#Feature and Label Extraction

X = train_df.drop('Item_Outlet_Sales', axis = 1)
Y = train_df['Item_Outlet_Sales']

In [50]:
X.head()

Item_Weight Item_Fat_Content  Item_Visibility              Item_Type  \
0         9.30               LF         0.016047                  Dairy   
1         5.92          Regular         0.019278            Soft Drinks   
2        17.50               LF         0.016760                   Meat   
3        19.20          Regular         0.000000  Fruits and Vegetables   
4         8.93               LF         0.000000              Household   

   Item_MRP  Outlet_Establishment_Year Outlet_Size Outlet_Location_Type  \
0  249.8092                       1999      Medium               Tier 1   
1   48.2692                       2009      Medium               Tier 3   
2  141.6180                       1999      Medium               Tier 1   
3  182.0950                       1998   Undefined               Tier 3   
4   53.8614                       1987        High               Tier 3   

         Outlet_Type  
0  Supermarket Type1  
1  Supermarket Type2  
2  Supermarket Type1  
3      Grocery Store  
4  Supermarket Type1

In [51]:
#Scaling those columns in which values differ greatly between their ranges
from sklearn.preprocessing import StandardScaler


cols_to_scale = ['Item_Weight', 'Item_Visibility','Item_MRP']
scaler = StandardScaler()
scaler.fit(X[cols_to_scale])
X[cols_to_scale] = scaler.transform(X[cols_to_scale])

In [54]:
X.head()

Item_Weight Item_Fat_Content  Item_Visibility              Item_Type  \
0    -0.209995               LF        -0.970732                  Dairy   
1    -0.735532          Regular        -0.908111            Soft Drinks   
2     1.064975               LF        -0.956917                   Meat   
3     1.329298          Regular        -1.281758  Fruits and Vegetables   
4    -0.267524               LF        -1.281758              Household   

   Item_MRP  Outlet_Establishment_Year Outlet_Size Outlet_Location_Type  \
0  1.747454                       1999      Medium               Tier 1   
1 -1.489023                       2009      Medium               Tier 3   
2  0.010040                       1999      Medium               Tier 1   
3  0.660050                       1998   Undefined               Tier 3   
4 -1.399220                       1987        High               Tier 3   

         Outlet_Type  
0  Supermarket Type1  
1  Supermarket Type2  
2  Supermarket Type1  
3      Grocery Store  
4  Supermarket Type1

In [59]:
##Applying Label Encoding to columns with categorical values

from sklearn.preprocessing import LabelEncoder

X[['Item_Fat_Content', 'Item_Type', 'Outlet_Establishment_Year', 'Outlet_Size', 'Outlet_Location_Type', 'Outlet_Type']] = X[['Item_Fat_Content', 'Item_Type', 'Outlet_Establishment_Year', 'Outlet_Size', 'Outlet_Location_Type', 'Outlet_Type']].apply(LabelEncoder().fit_transform)

In [60]:
X.head()

Item_Weight  Item_Fat_Content  Item_Visibility  Item_Type  Item_MRP  \
0    -0.209995                 0        -0.970732          4  1.747454   
1    -0.735532                 2        -0.908111         14 -1.489023   
2     1.064975                 0        -0.956917         10  0.010040   
3     1.329298                 2        -1.281758          6  0.660050   
4    -0.267524                 0        -1.281758          9 -1.399220   

   Outlet_Establishment_Year  Outlet_Size  Outlet_Location_Type  Outlet_Type  
0                          4            1                     0            1  
1                          8            1                     2            2  
2                          4            1                     0            1  
3                          3            3                     2            0  
4                          1            0                     2            1

# Preprocessing on Test Data

In [61]:
test_df.head()

Item_Weight Item_Fat_Content  Item_Visibility    Item_Type  Item_MRP  \
0       20.750               LF         0.007565  Snack Foods  107.8622   
1        8.300              Reg         0.038428        Dairy   87.3198   
2       14.600               LF         0.099575       Others  241.7538   
3        7.315               LF         0.015388  Snack Foods  155.0340   
4        0.000          Regular         0.118599        Dairy  234.2300   

   Outlet_Establishment_Year Outlet_Size Outlet_Location_Type  \
0                       1999      Medium               Tier 1   
1                       2007   Undefined               Tier 2   
2                       1998   Undefined               Tier 3   
3                       2007   Undefined               Tier 2   
4                       1985      Medium               Tier 3   

         Outlet_Type  
0  Supermarket Type1  
1  Supermarket Type1  
2      Grocery Store  
3  Supermarket Type1  
4  Supermarket Type3

In [62]:
#Scaling those columns in which values differ greatly between their ranges
from sklearn.preprocessing import StandardScaler


cols_to_scale = ['Item_Weight', 'Item_Visibility','Item_MRP']
scaler = StandardScaler()
scaler.fit(test_df[cols_to_scale])
test_df[cols_to_scale] = scaler.transform(test_df[cols_to_scale])

In [63]:
test_df.head()

Item_Weight Item_Fat_Content  Item_Visibility    Item_Type  Item_MRP  \
0     1.599453               LF        -1.134077  Snack Foods -0.536555   
1    -0.346052              Reg        -0.531850        Dairy -0.868937   
2     0.638421               LF         0.661316       Others  1.629848   
3    -0.499974               LF        -0.981416  Snack Foods  0.226697   
4    -1.643056          Regular         1.032540        Dairy  1.508110   

   Outlet_Establishment_Year Outlet_Size Outlet_Location_Type  \
0                       1999      Medium               Tier 1   
1                       2007   Undefined               Tier 2   
2                       1998   Undefined               Tier 3   
3                       2007   Undefined               Tier 2   
4                       1985      Medium               Tier 3   

         Outlet_Type  
0  Supermarket Type1  
1  Supermarket Type1  
2      Grocery Store  
3  Supermarket Type1  
4  Supermarket Type3

In [64]:
##Applying Label Encoding to columns with categorical values

from sklearn.preprocessing import LabelEncoder

test_df[['Item_Fat_Content', 'Item_Type', 'Outlet_Establishment_Year', 'Outlet_Size', 'Outlet_Location_Type', 'Outlet_Type']] = test_df[['Item_Fat_Content', 'Item_Type', 'Outlet_Establishment_Year', 'Outlet_Size', 'Outlet_Location_Type', 'Outlet_Type']].apply(LabelEncoder().fit_transform)

In [65]:
test_df.head()

Item_Weight  Item_Fat_Content  Item_Visibility  Item_Type  Item_MRP  \
0     1.599453                 0        -1.134077         13 -0.536555   
1    -0.346052                 1        -0.531850          4 -0.868937   
2     0.638421                 0         0.661316         11  1.629848   
3    -0.499974                 0        -0.981416         13  0.226697   
4    -1.643056                 2         1.032540          4  1.508110   

   Outlet_Establishment_Year  Outlet_Size  Outlet_Location_Type  Outlet_Type  
0                          4            1                     0            1  
1                          7            3                     1            1  
2                          3            3                     2            0  
3                          7            3                     1            1  
4                          0            1                     2            3

# Modelling

In [66]:
#train test split
from sklearn.model_selection import train_test_split
X_train, X_test, Y_train, Y_test = train_test_split(X,Y, test_size=0.20, random_state=42)

In [69]:
from sklearn.linear_model import LinearRegression
model = LinearRegression()
model.fit(X_train, Y_train)
Y_pred = model.predict(X_test)

In [71]:

from sklearn.metrics import mean_squared_error
mean_squared_error(Y_test,Y_pred)


1281720.140280301

In [74]:
from sklearn.neighbors import KNeighborsRegressor

KN = KNeighborsRegressor(n_neighbors=10)
KN.fit(X_train, Y_train)
Y_pred_KN = KN.predict(X_test)

mean_squared_error(Y_test, Y_pred_KN)



1212452.0339378824

In [76]:
from sklearn.ensemble import RandomForestRegressor

RF = RandomForestRegressor(random_state=45,max_depth=16,n_estimators=200)
RF.fit(X_train, Y_train)
Y_pred_RF = RF.predict(X_test)

mean_squared_error(Y_test, Y_pred_RF)
#print('{:0.2f}'.format(mse_rf))

1155505.4921311522

After modeeling with three models, LinearRegression, KNeighborsRegressor and RandomForestRegressor; RandomForestRegressor gives the least mse - 1155505.4921311522

So, going forward with this model and applying fine tuning.

Fine Tuning with Hyperpot

In [77]:
#best possible RandomForestRegressor model using hyperopt

from hyperopt import tpe, hp, fmin, STATUS_OK,Trials

In [78]:
space ={'n_estimators':hp.uniform('n_estimators',100,500),
           'max_depth':hp.uniform('max_depth',5,20),
           'min_samples_leaf':hp.uniform('min_samples_leaf',1,5),
           'min_samples_split':hp.uniform('min_samples_split',2,6)}

In [81]:
# define objective function

from sklearn.model_selection import cross_val_score

def hyperparameter_tuning(params):
    clf = RandomForestRegressor(**params,n_jobs=-1)
    acc = mean_squared_error(Y_test, Y_pred_RF).mean()
    return {"loss": -acc, "status": STATUS_OK}

In [82]:
# Initialize trials object
trials = Trials()

best = fmin(
    fn=hyperparameter_tuning,
    space = space, 
    algo=tpe.suggest, 
    max_evals=50, 
    trials=trials
)

print("Best: {}".format(best))

100%|██████████| 50/50 [00:00<00:00, 78.79it/s, best loss: -1155505.4921311522]
Best: {'max_depth': 7.185623800975174, 'min_samples_leaf': 3.233576246059008, 'min_samples_split': 2.3335796514722666, 'n_estimators': 189.53971726750885}


# Best Fit after Hyperparameter Tuning

In [84]:

RF = RandomForestRegressor(random_state=45, max_depth=7, n_estimators=189, min_samples_split=2, min_samples_leaf=3)
RF.fit(X_train, Y_train)
Y_pred_RF = RF.predict(X_test)

mse_RF = mean_squared_error(Y_test, Y_pred_RF)
print('{:0.2f}'.format(mse_RF))

1051241.45


Best fit RandomForestRegressor Model After Hyperparameter Tuning with a MSE of 1051241.45

# Test data Prediction using Best Fit Model

In [85]:
Test_Prediction = RF.predict(test_df)
Test_Prediction

array([1629.16064955, 1361.36097138,  570.03191385, ..., 1921.60485276,
       3609.87532713, 1290.03488211])

# Creating Submission csv File

In [87]:
submission = pd.DataFrame()


In [113]:
submission[["Item_Identifier","Outlet_Identifier"]] = test_df[["Item_Identifier","Outlet_Identifier"]]
submission["Item_Outlet_Sales"] = abs(Test_Prediction)

In [114]:
submission

Item_Identifier Outlet_Identifier  Item_Outlet_Sales
0              FDW58            OUT049        1629.160650
1              FDW14            OUT017        1361.360971
2              NCN55            OUT010         570.031914
3              FDQ58            OUT017        2447.041775
4              FDY38            OUT027        6426.843114
...              ...               ...                ...
5676           FDB58            OUT046        2127.746909
5677           FDD47            OUT018        2694.454375
5678           NCO17            OUT045        1921.604853
5679           FDJ26            OUT017        3609.875327
5680           FDU37            OUT045        1290.034882

[5681 rows x 3 columns]

In [115]:
!pwd

/content


In [116]:
submission.to_csv("submission.csv",index = False) #creating csv file

Submitted By: Ajuma Mohammed